wal-vdom/src/virtual_dom/vnode.rs

```diff
@@ -1,24 +1,172 @@
-use gloo::console::__macro::JsValue;
+use gloo::{console::log, utils::document};
 use serde::Serialize;
-use web_sys::Node;
+use web_sys::{Element, Node, Text};
 
 use super::{VElement, VText};
 
-#[derive(Serialize)]
+#[derive(Debug, Serialize)]
 pub enum VNode {
-    Element{
-        velement: VElement,
+    Element {
+        #[serde(skip_serializing)]
+        concrete: Option<Element>,
+        virt: VElement,
     },
-    Text{
-        vtext: VText,
+    Text {
+        #[serde(skip_serializing)]
+        concrete: Option<Text>,
+        virt: VText,
     },
 }
 
 impl VNode {
-    pub fn render(&self) -> Result<Node, JsValue> {
+    pub fn patch(&mut self, last: Option<VNode>, ancestor: &Node) {
         match self {
-            VNode::Element { velement } => velement.render(),
-            VNode::Text { vtext } => vtext.render(),
+            VNode::Element {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_element(concrete, virt, last, ancestor),
+            VNode::Text {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_text(concrete, virt, last, ancestor),
+        };
+    }
+
+    fn patch_element(
+        concrete: &mut Option<Element>,
```

Comment:
czy concrete i last nie są mocno powiązane? Bo concrete to jest ostatni element stworzony z virtual noda last. Jeśli tak można by było jakoś połączyć to w jednego structa i by to uprościło kod. No chyba że jest możliwe że last to Some a concrete to None.